Source/DafnyCore/Substituter.cs

```diff
@@ -38,13 +38,19 @@ public class Substituter {
         return new StaticReceiverExpr(e.tok, ty, e.IsImplicit) { Type = ty };
       } else if (expr is LiteralExpr literalExpr) {
         if (literalExpr.Value == null) {
-          return new LiteralExpr(literalExpr.tok) { Type = Resolver.SubstType(literalExpr.Type, typeMap) };
+          var ty = Resolver.SubstType(literalExpr.Type, typeMap);
```

Comment:
There's a (possibly misguided) optimization in the `Substitute(Expression expr)` method, which is that if the result is pointer-equal to `expr` if the substitution has no effect. An illustrative example of this is:

``` c#
      } else if (expr is MultiSetFormingExpr) {
        var e = (MultiSetFormingExpr)expr;
        var se = Substitute(e.E);
        if (se != e.E) {
          newExpr = new MultiSetFormingExpr(expr.tok, se);
        }
```

This optimization is error prone. For `LiteralExpr`, the optimization was missing, so I added it. In other places (like `DatatypeValue`, the optimization was too aggressive (meaning, "wrong") because it returned `expr` even when it should have returned a new `DatatypeValue` with a different `InferredTypeArgs` field.

The optimization is intended to save memory, but this may not make any difference, and the optimization makes the code significantly more complicated. If you suggest, I'd be happy to remove this optimization from `Substitute(Expression`.

Reply:
I built a local version with a simplified `Substitute` (that is, I removed the optimization). It passes. I'd be happy to add it to this PR, if you think, or to file a separate PR for it.

Reply:
> The optimization is intended to save memory, but this may not make any difference, and the optimization makes the code significantly more complicated.

Without a measurement about the performance difference I'd prefer to leave it like it is, unless you're confident the optimisation doesn't have a significant effect. I'm not sure whether comparing the difference in the test-suite runtime is a good test.

I'm hoping we can at some point generate the code for various behaviors such as Substitutor, Cloner, SubExpressions, SubStatements, and likely some others.

Reply:
Okay, I'll leave it as is. And yes, I agree that _generating_ code for these sorts of routines would be far better than the manual upkeep we have to do today.